volume/drivers/pwx/connection.go

```diff
@@ -0,0 +1,210 @@
+	// as environment variables
+	serviceName := getPxService(cpb.Config.ServiceNameEnv, cpb.Config.DefaultServiceName)
+	ns := getPxNamespace(cpb.Config.NamespaceNameEnv, cpb.Config.DefaultServiceNamespaceName)
+
+	svc, err := cpb.kubeOps.GetService(serviceName, ns)
+	if err == nil {
+		endpoint = svc.Spec.ClusterIP
+	} else {
+		return "", "", fmt.Errorf("failed to get k8s service spec: %v", err)
+	}
+
+	if len(endpoint) == 0 {
+		return "", "", fmt.Errorf("failed to get endpoint for portworx volume driver")
+	}
+
+	var restPort int
+	var sdkPort int
+
+	// Get the ports from service
+	for _, svcPort := range svc.Spec.Ports {
+		if svcPort.Name == cpb.Config.DefaultSDKPortName &&
+			svcPort.Port != 0 {
+			sdkPort = int(svcPort.Port)
+		} else if svcPort.Name == cpb.Config.DefaultRestPortName &&
+			svcPort.Port != 0 {
+			restPort = int(svcPort.Port)
+		}
+	}
+
+	// check if the ports were parsed
+	if sdkPort == 0 || restPort == 0 {
+		err := fmt.Errorf("%s in %s namespace has been not updated to the latest spec. "+
+			"%s and/or %s ports are missing. Follow "+
+			"https://docs.portworx.com/portworx-install-with-kubernetes/operate-and-maintain-on-kubernetes/upgrade/"+
+			" to upgrade Portworx", serviceName, ns, cpb.Config.DefaultSDKPortName, cpb.Config.DefaultRestPortName)
+		logrus.Errorf(err.Error())
+		return "", "", err
+	}
+
+	// PX mgmt API was decided not to be protected with TLS
+	pxMgmtEndpoint = fmt.Sprintf("http://%v:%v", endpoint, restPort)
+	sdkEndpoint = fmt.Sprintf("%s:%d", endpoint, sdkPort)
+
+	logrus.Infof("Using %s as endpoint for portworx REST endpoint", pxMgmtEndpoint)
+	logrus.Infof("Using %s as endpoint for portworx gRPC endpoint", sdkEndpoint)
+
+	return pxMgmtEndpoint, sdkEndpoint, nil
+}
+
+// BuildDialOps build slice of grpc.DialOption to connect to SDK API
+func (cpb *ConnectionParamsBuilder) BuildDialOps() ([]grpc.DialOption, error) {
+	if !isTLSEnabled(cpb.Config.EnableTLSEnv) {
+		return []grpc.DialOption{grpc.WithInsecure()}, nil
+	}
+
+	var rootCA []byte
+	var caCertSecretName = strings.TrimSpace(os.Getenv(cpb.Config.CaCertSecretEnv))
+	var caCertSecretKey = strings.TrimSpace(os.Getenv(cpb.Config.CaCertSecretKeyEnv))
+	var pxNamespace = getPxNamespace(cpb.Config.NamespaceNameEnv, cpb.Config.DefaultServiceNamespaceName)
+
+	if caCertSecretName != "" && caCertSecretKey == "" {
+		return nil, fmt.Errorf("failed to load CA cert from secret: %s, secret key should be defined using env PX_CA_CERT_SECRET_KEY", caCertSecretName)
+	}
+
+	if caCertSecretName != "" && caCertSecretKey != "" {
+		secret, err := cpb.kubeOps.GetSecret(caCertSecretName, pxNamespace)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load CA cert secret: %v", err)
+		}
+
+		exist := false
+		rootCA, exist = secret.Data[caCertSecretKey]
+		if !exist {
+			return nil, fmt.Errorf("failed to load CA cert from secret: %s using key: %s", caCertSecretName, caCertSecretKey)
+		}
+		if len(rootCA) == 0 {
```

Comment:
Does this mean if the caller calls this function there _must_ be a rootCA? What if the CA is already part of the container/host?

Reply:
I add that check intentionally. My reasoning is: if a user has configured service to read CA cert from a specific secret/secret key we should expect a certificate.
If CA cert is part of container already a user should just enable TLS and do not configure an application to read CA cert from k8s secret.